site/app/controllers/grading/ElectronicGraderController.php

```diff
@@ -1283,6 +1283,144 @@ private function amIBlindGrading($gradeable, $user, $peer) {
+        // check for rollback submission only if the Active version is greater than 1 and that too is late.
+        if ($previousVersion && $late_status !== LateDayInfo::STATUS_GOOD) {
+            while ($previousVersion) {
+                $prevVersionInstance = $graded_gradeable->getAutoGradedGradeable()->getAutoGradedVersionInstance($previousVersion);
+                if ($prevVersionInstance == null) {
+                    $rollbackSubmission = -1;
+                    break;
+                }
+                $lateInfo = LateDays::fromUser($this->core, $late_days_user)->getLateDayInfoByGradeable($gradeable);
+                $daysLate = $prevVersionInstance->getDaysLate();
+
+                // If this version is a good submission then it the rollback Submision
+                if ($lateInfo == null || ($lateInfo->getStatus($daysLate) == LateDayInfo::STATUS_GOOD)) {
+                    $rollbackSubmission = $previousVersion;
+                    break;
+                }
+                // applying same condition for previous version. i.e going back one version
+                $previousVersion -= 1;
+            }
+        }
+
+        $logger_params = [
+            "course_semester" => $this->core->getConfig()->getSemester(),
+            "course_name" => $this->core->getDisplayedCourseName(),
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "submitter_id" => $who_id,
+            "action" => "VIEW_PAGE",
+        ];
+        Logger::logTAGrading($logger_params);
+        $anon_mode = isset($_COOKIE['anon_mode']) && $_COOKIE['anon_mode'] === 'on';
+        $submitter_itempool_map = $this->getItempoolMapForSubmitter($gradeable, $graded_gradeable->getSubmitter()->getId());
+        $solution_ta_notes = $this->getSolutionTaNotesForGradeable($gradeable, $submitter_itempool_map) ?? [];
+
+        $this->core->getOutput()->addInternalCss('forum.css');
+        $this->core->getOutput()->addInternalCss('electronic.css');
+
+        $this->core->getOutput()->addInternalJs('forum.js');
+        $this->core->getOutput()->addInternalCss('grade-inquiry.css');
+        $this->core->getOutput()->addInternalJs('grade-inquiry.js');
+        $this->core->getOutput()->addInternalJs('websocket.js');
+        $show_hidden = $this->core->getAccess()->canI("autograding.show_hidden_cases", ["gradeable" => $gradeable]);
+        $this->core->getOutput()->renderOutput(['grading', 'ElectronicGrader'], 'hwGradingPopup', $gradeable, $graded_gradeable, $panel, $display_version, $show_hidden, $can_inquiry, $can_verify, $show_verify_all, $show_silent_edit, $rollbackSubmission, $who_id, $solution_ta_notes, $submitter_itempool_map, $anon_mode, $blind_grading, $window_id);
+        $this->core->getOutput()->renderOutput(['grading', 'ElectronicGrader'], 'popupStudents');
+        $this->core->getOutput()->renderOutput(['grading', 'ElectronicGrader'], 'popupMarkConflicts');
```

Comment:
I'm not too familiar with this part of the codebase so there may be no better way to do it without significant changes but ideally this type of code would be moved to a view instead of a controller.